src/machine/io.c

```diff
@@ -19,45 +19,79 @@
 #include <stdint.h>
 
 /*
- * a handle defining how to output a character
+ *------------------------------------------------------------------------------
+ * printf() output channel management
+ *------------------------------------------------------------------------------
  */
-typedef void (*out_fn)(char character, char *buf, word_t idx);
 
-/*
- * structure to allow a generic vprintf
- * a out_fn handle and a buf to work on
- */
-typedef struct {
-    out_fn putchar;
-    char *buf;
-    word_t idx;
-    word_t maxlen;
-} out_wrap_t;
+typedef struct _out_wrap_t  out_wrap_t;
 
-/*
- * putchar would then just call the handle with its buf
- * and current idx and then increment idx
+/* handler defining how/where to actually output a buffer */
+typedef void (*out_write_fn)(out_wrap_t *out, const char *buf, word_t len);
+
+struct _out_wrap_t {
```

Comment:
Can you expand the commit message for this change to be more descriptive? It's changing the interface more than what a clean-up would indicate.